cluster/v1alpha1/helpers.go

```diff
@@ -99,171 +98,179 @@ func (r *RolloutHandler) GetRolloutCluster(rolloutStrategy RolloutStrategy, stat
 	}
 
-	// upgrade mandatory decision groups first
-	groupKeys := decisionGroupsToGroupKeys(strategy.ProgressivePerGroup.MandatoryDecisionGroups.MandatoryDecisionGroups)
+	// Upgrade mandatory decision groups first
+	mandatoryDecisionGroups := strategy.ProgressivePerGroup.MandatoryDecisionGroups.MandatoryDecisionGroups
+	groupKeys := decisionGroupsToGroupKeys(mandatoryDecisionGroups)
 	clusterGroupKeys, clusterGroups := r.pdTracker.ExistingClusterGroups(groupKeys)
 
-	rolloutresult := progressivePerGroup(clusterGroupKeys, clusterGroups, maxTimeDuration, statusFunc)
-	if len(rolloutresult.ClustersToRollout) > 0 {
-		return &strategy, rolloutresult, nil
+	// Perform progressive rollout per group for mandatory decision groups
+	rolloutResult := progressivePerGroup(clusterGroupKeys, clusterGroups, maxTimeDuration, statusFunc)
+	if len(rolloutResult.ClustersToRollout) > 0 {
+		return &strategy, rolloutResult, nil
 	}
 
-	// parse timeout
+	// Parse timeout for non-mandatory decision groups
 	failureTimeout, err := parseTimeout(strategy.ProgressivePerGroup.Timeout.Timeout)
 	if err != nil {
 		return &strategy, RolloutResult{}, err
 	}
 
-	// upgrade the rest decision groups
+	// Upgrade the rest of the decision groups
 	restClusterGroupKeys, restClusterGroups := r.pdTracker.ExistingClusterGroupsBesides(clusterGroupKeys)
 
-	rolloutresult = progressivePerGroup(restClusterGroupKeys, restClusterGroups, failureTimeout, statusFunc)
-	return &strategy, rolloutresult, nil
+	// Perform progressive rollout per group for the remaining decision groups
+	rolloutResult = progressivePerGroup(restClusterGroupKeys, restClusterGroups, failureTimeout, statusFunc)
+	return &strategy, rolloutResult, nil
 }
 
 func progressivePerCluster(clusters []string, clusterToGroupKey map[string]clusterv1beta1.GroupKey, length int, timeout time.Duration, statusFunc ClusterRolloutStatusFunc) RolloutResult {
-	rolloutclusters := map[string]ClusterRolloutStatus{}
-	timeoutclusters := map[string]ClusterRolloutStatus{}
+	rolloutClusters := map[string]ClusterRolloutStatus{}
+	timeoutClusters := map[string]ClusterRolloutStatus{}
 
 	if length == 0 {
 		return RolloutResult{
-			ClustersToRollout: rolloutclusters,
-			ClustersTimeOut:   timeoutclusters,
+			ClustersToRollout: rolloutClusters,
+			ClustersTimeOut:   timeoutClusters,
 		}
 	}
 
-	// sort the clusters in alphabetical order, ensure each time returns the same clusters.
+	// Sort the clusters in alphabetical order to ensure consistency.
 	sort.Strings(clusters)
 	for _, cluster := range clusters {
 		status := statusFunc(cluster)
-		if groupKey, exist := clusterToGroupKey[cluster]; exist {
+		if groupKey, exists := clusterToGroupKey[cluster]; exists {
 			status.GroupKey = groupKey
 		}
 
-		newstatus, needtorollout := needToRollout(status, timeout)
-		status.Status = newstatus.Status
-		status.TimeOutTime = newstatus.TimeOutTime
+		newStatus, needToRollout := needToRollout(status, timeout)
```

Comment:
It's confusing that `needToRollout` is the same name as the `needToRollout` function.

Reply:
Update to `newStatus, needToRollout := determineRolloutStatusAndContinue(status, timeout)`